improve error handling on failed Http/2 handshake

wfurt opened the pull request:
The issue described in #36665 was misconfiguration. However getting ObjectDisposedException on internal objects is not very helpful. 

The RFC says:
> The server connection preface consists of a potentially empty  SETTINGS frame (Section 6.5) that MUST be the first frame the server  sends in the HTTP/2 connection.

This change modifies SetupAsync() to check for server preface and throw error immediately instead of aborting connection in background task. It will also log few bytes from response if tracing is enable so it is easier to diagnose this in field. 

with this called would get following exception instead of ObjectDisposedException:
```
System.Net.Http.HttpRequestException: The server returned an invalid or unrecognized response.
   at System.Net.Http.Http2Connection.SetupAsync()
   at System.Net.Http.HttpConnectionPool.GetHttp2ConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at proxy_test.Program.fetch2(String uri) in /home/furt/proxy-test/Program.cs:line 58
```
closes #36665

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -138,6 +138,25 @@ public async Task SetupAsync()
+            // Receive the initial SETTINGS frame from the peer.
+            try
+            {
+                FrameHeader frameHeader = await ReadFrameAsync(initialFrame : true).ConfigureAwait(false);
```
Nit: unnecessary space before colon

davidsh commented:
There are no test changes/additions in this PR.  Is this difficult to have automated tests?  How did you test these changes?

wfurt commented:
I set up HTTP 1.1 server and I used ssl proxy in front of it advertising HTTP2 via ALPN. So HttpClient would fall to HTTP2 processing while server does not understand it. 

After second thought I modified  old Loopback server and added test for this.

davidsh commented on src/Common/tests/System/Net/Http/LoopbackServer.cs:
```diff
@@ -355,6 +361,7 @@ public class Options
                 SslProtocols.Tls13 |
 #endif
                 SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
+            public List<SslApplicationProtocol> ApplicationProtocols;
```
This should not be a public field.  This is against usual .NET design guidelines.

Instead, it should be a public property with a getter only to return a list.  The list should be initialized to an empty list.

Then consumers can simply do this:

```c#
Options options = new Options();
options.ApplicationProtocols.Add(SslApplicationProtocol.Http2);
```

wfurt replied:
I had to roll this back as that broke tests outside of System.Net.Http built for netstandard(2.0)

wfurt commented:
/azp run corefx-outerloop-linux

wfurt commented:
/azp run corefx-outerloop-osx

wfurt commented:
/azp run corefx-outerloop-windows

wfurt commented:
outerloop failures seems unrelated.

davidsh commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -138,6 +138,25 @@ public async Task SetupAsync()
+            catch
+            {
+                // ISSUE 31315: Determine if/how to expose HTTP2 error codes
+                throw new HttpRequestException(SR.net_http_invalid_response);
```
We typically would insert the the exception from the 'catch' into the HttpRequestException.  Otherwise, it is difficult to diagnose anything. So, this should be changed to something like:

```c#
catch (Exception ex)
{
    throw new HttpRequestException(SR.net_http_invalid_response, ex);
```

wfurt replied:
updated.

geoffkizer commented:
If I understand correctly, this change means we will wait to receive the initial SETTINGS from the server before trying to actually use the connection to send requests. Correct?

Per spec, we are allowed to start sending requests before we receive the initial SETTINGS from the server. Forcing us to wait for the server SETTINGS may introduce a perf issue here.

I'm not sure why we can't keep the SETTINGS handling on the background task, and just handle the exception better?

wfurt commented:
We do send frames out before this in SetupAsync() so we are not waiting. Do you have specific recommendation how to get exceptions back? There was recent discussion about sending request in parallel with receiving response for gRPC and it will be useful there as well.

geoffkizer commented:
> We do send frames out before this in SetupAsync() so we are not waiting.

These are just the connection establishment frames though, right? I'm talking about sending actual requests.

> Do you have specific recommendation how to get exceptions back?

I'm not sure what you are asking. What exactly are we trying to fix here? There are lots of ways the connection can fail once it's established, why is this particular case any different?

wfurt commented:
I have two goals: 
1. collect better (more detailed?) diagnostic on failure on first frame as that is likely may be protocol mismatch
2. throw better exaction to caller (like The server returned an invalid or unrecognized response.) instead of ObjectDisposed on internal objects.

Aside from the CreditManager, I have also seen disposed exceptions on SslStream.

wfurt commented:
I updated code to avoid delay. Now we will throw:

```
        System.Net.Http.HttpRequestException : The server returned an invalid or unrecognized response.
        ---- System.Net.Http.Http2Connection+Http2ProtocolException : Http2 Protocol Error, errorCode = ProtocolError
        Stack Trace:
          /home/furt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs(1381,0): at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs(525,0): at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs(33,0): at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs(507,0): at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs(135,0): at System.Net.Http.HttpClient.GetStringAsyncCore(Task`1 getTask)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(1300,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.<>c__DisplayClass47_0.<<Http2_ProtocolMismatch_Throws>b__0>d.MoveNext()
          --- End of stack trace from previous location where exception was thrown ---
          /home/furt/github/wfurt-corefx/src/Common/tests/System/Net/Http/LoopbackServer.cs(66,0): at System.Net.Test.Common.LoopbackServer.CreateServerAsync(Func`2 funcAsync, Options options)
          /home/furt/github/wfurt-corefx/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs(1282,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.Http2_ProtocolMismatch_Throws()
          --- End of stack trace from previous location where exception was thrown ---
```

wfurt commented:
I sync up with #37223. However even with that change, _abortException did not bubble up and we were still getting  ObjectDisposedException on CreditManager. I made some updates as well as I updated new test to check for type of InnerException as everything now gets wrapped in HttpRequestException.

stephentoub commented:
> _abortException did not bubble up

Where didn't it bubble up?

wfurt commented:
to caller of client.GetStringAsync() e.g. test in this case:
```
  System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.ObjectDisposedException: Cannot access a disposed object.
  Object name: 'CreditManager'.
     at System.Net.Http.CreditManager.RequestCreditAsync(Int32 amount, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:line 50
     at System.Net.Http.Http2Connection.SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 965
     at System.Threading.Tasks.ValueTask`1.get_Result() in /__w/1/s/src/System.Private.CoreLib/shared/System/Threading/Tasks/ValueTask.cs:line 759
     at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 1353
     --- End of inner exception stack trace ---
     at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:line 1393
     at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 525
     at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs:line 33
     at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 507
     at System.Net.Http.HttpClient.GetStringAsyncCore(Task`1 getTask) in /mnt/github/wfurt-corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 135
     at System.Net.Http.Functional.Tests.HttpClientHandlerTest_Http2.<>c__DisplayClass49_0.<<Http2_ProtocolMismatch_Throws>b__0>d.MoveNext() in /mnt/github/wfurt-corefx/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs:line 1349
      System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2.Http2_ProtocolMismatch_Throws [FAIL]
```

This is because the inner exception does not come through Http2Stream.

stephentoub commented:
> to caller of client.GetStringAsync() e.g. test in this case:

Ah.  My PR only changed call sites to throw _abortException when they were checking for the aborted state; this doesn't appear to be doing that.

wfurt commented:
The stream it self should be in aborted state but the exception comes directly from `Http2Connection.SendHeadersAsync()`. Since the _abortException is same for all streams, could we store it just in  Http2Connection and use _connection. _abortException in Http2Stream code? 
When I saw your PR I would not sure if there is benefit of putting it each stream.

stephentoub commented:
> Since the _abortException is same for all streams

It's not; different streams could have different abort exceptions due to an RST_STREAM frame specific to that stream.

wfurt commented:
/azp run

wfurt commented:
That make sense @stephentoub . I look at it again and proposed change should work as intended as it brings connection-level errors from Abort()   

outerloop failures are not related to this PR.

wfurt commented:
/azp run corefx-outerloop-windows

wfurt commented:
/azp run corefx-outerloop-linux

wfurt commented:
/azp run corefx-outerloop-osx

wfurt commented:
outerloop failures seems unrelated.

ViktorHofer commented:
@wfurt why did you merge this PR when all legs were failing? because of a build error?

```
Build FAILED.

HttpClientHandlerTest.Http2.cs(1352,10): error CS1513: } expected [F:\workspace\_work\1\s\src\System.Net.Http\tests\FunctionalTests\System.Net.Http.Functional.Tests.csproj]
    0 Warning(s)
    1 Error(s)
```

https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs#L1351 is missing a bracket:

ViktorHofer commented:
I commited directly into master to fix the build break: https://github.com/dotnet/corefx/commit/3075793b035cfa51d5b2a8aa0d388ee41faf2bf9

wfurt commented:
sorry @ViktorHofer There was merge conflict I tried to resolve using github UI. and obviously I did not do it quite right.
